lib/dry/operation/class_context.rb

```diff
@@ -0,0 +1,82 @@
+# frozen_string_literal: true
+
+module Dry
+  class Operation
+    # {Dry::Operation} class context
+    module ClassContext
+      # Default methods to be prepended unless changed via {.operate_on}
+      DEFAULT_METHODS_TO_PREPEND = [:call].freeze
+
+      # Configures the instance methods to be prepended
+      #
+      # The given methods will be prepended with a wrapper that calls {#steps}
+      # before calling the original method.
+      #
+      # This method must be called before defining any of the methods to be
+      # prepended or before prepending any other method.
+      #
+      # @param methods [Array<Symbol>] methods to prepend
+      # @raise [MethodsToPrependAlreadyDefinedError] if any of the methods have
+      #  already been defined in self
+      # @raise [PrependConfigurationError] if there's already a prepended method
+      def operate_on(*methods)
+        @_mutex.synchronize do
+          @_prepend_manager = @_prepend_manager.register(*methods)
+        end
```

Comment:
Could we avoid having to deal with these mutexes if we just allowed the prepend manager to mutate itself? i.e. turn `#register` and `#void` into methods that mutate the manager instead of returning a new copy.

We would continue to assign a copy of the manager upon inheritance (see `.indirectly_inherited`), which will ensure there will be no changes leaking across class boundaries, too.

Reply:
Not having to create all those instances is probably nicer for perf too.

Reply:
> we could just use #dup at the time of inheritance.

I'm afraid that wouldn't work, as [`#dup`](https://rubyapi.org/3.2/o/object#method-i-dup) won't copy the internal `methods_to_prepend` & `prepended_method` arrays, so any change to them in any of the `PrependerManager` instances would affect all of them:

```ruby
class A
  attr_reader :foo
  
  def initialize
    @foo = []
  end

  def push(value)
    @foo << value
  end
end

a = A.new
a.push(:bar)
a.foo # => [:bar]
b = a.dup
b.push(:var)
b.foo # => [:bar, :var]
a.foo # => [:bar, :var]
```

We could define a custom implementation of `#dup`, or `#dup` the hashes internally on the `PrependManager` methods, but considering all of this it looks to me that swapping the whole instance is the cleaner way to manage the state here. What do you think? :thinking:

Reply:
Ah yes, I should have been more explicit, I really meant "use `#dup` and an appropriate `#initialize_copy` implementation". It'd be something like this:

```ruby
def initialize_copy(source)
  super

  @methods_to_prepend = source.methods_to_prepend.dup
  @prepended_methods = source.prepended_methods.dup
end
```

This would allow the `prepend_manager.dup` in the `.inherited` hook to work as expected.

Reply:
I ended up adding a `#for_subclass` method to the prepend manager, as we also need to swap the `@klass` instance variable. It also encapsulates better the behavior. As you suggested, the prepend manager controls now its own state. Please, take a look at the last commit. If you agree with the implementation I'll squash it into the previous one :slightly_smiling_face: